enhancements/installer/aws-custom-edge-machineset-local-zones.md

```diff
@@ -0,0 +1,677 @@
+  compute pool, and keep it dedicated for applications which are latency sensitive,
+  so I can have a cost-effective infrastructure, avoiding mixing workloads in different locations.
+
+- As a cluster admin, I would like to select existing subnets from the local and
+  the parent region zones, to install a cluster, so I can manage my custom networks with
+  my automation.
+
+Phase 2:
+
+- As a cluster admin, I would like to install OpenShift clusters, extending the
+  compute nodes to the Local Zones in my day zero operations without needing to
+  set up the network and compute dependencies, so I can speed up the edge adoption
+  in my organization using OKD/OCP.
+
+### Goals
+
+Phase 1:
+
+- Admins can create compute pool named `edge` on the AWS platform to setting up Local Zone MachineSets.
+- Admins can select and configure subnets on Local Zones before cluster creation.
+- Ensure the installer allows creating a new machine pool for `edge` workloads
+- Ensure the installer can create the MachineSet with `NoSchedule` taints on edge machine pools.
+- Ensure Local Zone subnets will not be used on `worker` compute pools or control planes.
+- Ensure the Wavelength zone will not be used in any compute pool
+
+Phase 2:
+
+- Admins can configure the Local availability zones before cluster creation.
+- Ensure the installer creates network resources based on the Local Zone defined on the
+  `edge` compute pool.
+
+### Non-Goals
+
+- Allow multiple compute pools for `edge`
+- Allow multiple instance types for the edge compute pool
+- Support private Local Zone subnets
+- Create any resource for Wavelength zones. Although Wavelength is also running on
+  the edge, it introduces a new resource called [Carrier Gateway](https://docs.aws.amazon.com/wavelength/latest/developerguide/carrier-gateways.html) to provide
+  connectivity between the compute nodes and the carrier network.
+
+Phase 1:
+
+- Create the network resources on the edge locations. That implementation will be
+  done in the second stage, when the installer will fully automate the creation
+  of resources on the availability zones specified on the `compute[.name=='edge'].platform.aws.zones`.
+
+## Proposal
+
+The phases proposed in this enhancement are:
+
+- Phase 1: Introduce the `edge` compute pool restricted to AWS Local Zones. Allow
+  users to install a cluster in existing networks with Local Zone subnets.
+- Phase 2: Create the network resources for AWS Local Zone locations.
+
+### Phase 1: Introduce the "edge" compute pool
+
+The first phase will introduce the new group of compute pool designed to
+be used exclusively for the user workloads on the zones not hosted in the
+default availability zones within the region, where the cluster is used to
+be deployed.
+
+Due to the limitation of cloud services in the edge locations, the dedicated
+pool also has limited compute and EBS type when compared to the default
+IPI (i.e. on AWS: m6i.xlarge).
```

Comment:
> Do these edge locations ever get newer instance types?

Not newest, most of them do not have the 6th-gen instances, currently the latest generation used by the installer.

> wondering if we will have similar issues with these edge zones where we will have to keep them up to date.

It will follow the same approach as nowadays. See my comment with regards to how the installer chooses the type based on it's [zone availability](https://github.com/openshift/enhancements/pull/1232/files#r979172587). The `edge` pool will use the same engine, but will not impact the decision of the current worker or master pool instance types.

> Is there a consistent set of instance types across all the edge zones that we can use as a default choice?

We can query the [instance offerings](https://aws.amazon.com/ec2/instance-types/) to check it. I exported it across locations between us-east-1 and us-west-2, take a look on the list:
https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42

As described on this EP, it seems we could use the existing installer preferred flow appending the `c5d.2xlarge` to cover all Local Zone locations, then the evaluation order will be: m6i.xlarge, m5.xlarge, and c5d.2xlarge.

Reply:
@JoelSpeed Thanks for your feedback. Revisiting this topic I can see one situation that could be one issue when defining a unique instance type for the entire pool (that's the only exception I can see):
- When deploying a cluster in `us-west-2` supporting `edge` pools on subnets `us-west-2-lax-1{a,b}` [Los Angeles] AND any other Local Zone location within `us-west-2` (let's use `us-west-2-las-1a` [Las Vegas])

With that scenario, the installer will lookup[1] for an intersection of instance types from the default Instances over **all AZs for the pool**. It will not find any, as `c5d.2xlarge` is not offered in any zone of Los Angeles, AND `m5.xlarge` is not offered in any zone out of Los Angeles. Finally, the default preferred type `m6.xlarge` will be choosed. It will result in the Failed machine when the cluster will be created. The Los Angeles location is the unique exception that does not offer `c5d.2xlarge` along all zones previously queried[2].

[1] https://github.com/openshift/installer/blob/master/pkg/asset/machines/aws/instance_types.go#L17-L39
[2] https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42

Said that, and considering that those pools will not run the same workload due to the taints, I propose to use the same preferred engine to choose the instance type to check the preferred order **by AZ**, as we are always creating MachineSet by AZ.

So in this case, given the new preferred list for those zones, the installer will generate those MachineSets:

```
>> 99_openshift-cluster-api_worker-machineset-0.yaml
replicas	: 1
AZ		: us-west-2a
instanceID	: m6i.xlarge
EBS		: gp3
zone_type	 null
meta		: {}

>> 99_openshift-cluster-api_worker-machineset-1.yaml
replicas	: 1
AZ		: us-west-2b
instanceID	: m6i.xlarge
EBS		: gp3
zone_type	 null
meta		: {}

>> 99_openshift-cluster-api_worker-machineset-2.yaml
replicas	: 1
AZ		: us-west-2c
instanceID	: m6i.xlarge
EBS		: gp3
zone_type	 null
meta		: {}

>> 99_openshift-cluster-api_worker-machineset-3.yaml
replicas	: 1
AZ		: us-west-2-lax-1a
instanceID	: m5.xlarge
EBS		: gp2
zone_type	 local-zone
meta		: {
  "labels": {
    "node-role.kubernetes.io/edge": "",
    "zone_group": "us-west-2-lax-1",
    "zone_type": "local-zone"
  }
}

>> 99_openshift-cluster-api_worker-machineset-4.yaml
replicas	: 1
AZ		: us-west-2-las-1a
instanceID	: c5d.2xlarge
EBS		: gp2
zone_type	 local-zone
meta		: {
  "labels": {
    "node-role.kubernetes.io/edge": "",
    "zone_group": "us-west-2-las-1",
    "zone_type": "local-zone"
  }
}

```

The "best/preferred" instances will be choosed chosen for each AZ: `m5.xlarge` for `us-west-2-lax-1a`, and `c5d.2xlarge` for `us-west-2-las-1a`

Thoughts?

Reply:
@JoelSpeed  could you please elaborate "we want to be more explicit about it"?

The logic proposed is to select automatically when the `compute[name==edge].platform.aws.type` will not be set, otherwise it will have more precedence. That strategy follows the current flow for `worker` machine pool. Is it make sense to you?

Reply:
Well I'm wondering if using this logic, there's a possibility that different worker zones might get different instance types and different edge zones may get different types from one another when they don't share.

I would expect the logic to go through each worker zone and check the available types and choose an instance type that is available across them all, and then repeat that process with the edge zones

Unless we are happy with edge foo being c5d.2xlarge and edge bar being m5.xlarge, would be odd for the same worker types to default to different instance types right?

Reply:
> (1) Well I'm wondering if using this logic, there's a possibility that different worker zones might get different instance types
> (2) and different edge zones may get different types from one another when they don't share.

(1) it will not happen. The current implementation will not be changed, preserving the current flow: the `worker` gets the same instance type across all regular zones.
(2) it will happen in this proposal. The proposal is to use the existing flow, get the preferred instance, looking to the scope of zone, instead of all zones, when choosing the `edge` instances. Example output on us-west-2 cluster deployment is available [here](https://github.com/openshift/installer/pull/6371#issuecomment-1260409687).

> I would expect the logic to go through each worker zone and check the available types and choose an instance type that is available across them all, and then repeat that process with the edge zones

Is it the idea with this logic of lookup of type offerings between all zones (regular and local) set the same instance between `worker` and `edge`?

If yes, I wonder if one of those problems will drive the solution more complex:
- the cluster costs will be increase (2x) as we will choose a common instance to use across all AZs: Currently, the intersection type between AZs is `c5d.2xlarge` (according to the [list exported above](https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42)), which costs 2x the `m6i.xlarge`
- the type `c5d.2xlarge` provides an ephemeral disk, which is good when there's a process to mount and use it by default. It's not the case with OCP, so we will not use this resource, which makes that instance type ~12% more expensive when compared with the same family without that storage (`c5.2xlarge`).
- we may need to mark some Local Zones as unused: `us-west-2-lax-1` zone group does not have "intersection" instance types between all zones. This zone was the first available on Local Zones, currently the unique zone group with two AZs in LZ, and the most modern one providing cheaper instances (`m5.xlarge`), and also `gp3` storage.
- we may change the entire cluster types when using a new feature (Local Zones/edge pool), which may increase the complexity in Support of changing the default instance types for worker and control plane nodes, instead of the expected `m6i.xlarge`.

> Unless we are happy with edge foo being c5d.2xlarge and edge bar being m5.xlarge, would be odd for the same worker types to default to different instance types right?

I am not sure if we would need to be concerned about different instance types between `worker` and `edge`, and within `edge` zone groups for those reasons:
- `worker` should not, initially, share the same workloads as `edge`
- nodes between different `edge` groups may not share the same workload. For that reason, we are setting taints and custom labels by zoneGroup to give control to admins/developers to choose what workloads to run for each edge location. For example, if the administrator creates a new ingress (ALB) sharing nodes from different Local Zones zone groups (i.e.: LAS and LAX), their end-users will have different experiences (sometimes worst) in terms of network latency.
- we are using a "best instance" strategy for the `edge` pools, decreasing the necessity of maintenance on the code
- we are still using the standard types from the default IPI (without edge/Local Zones support)

@JoelSpeed Thoughts? What do you think about those points?

Reply:
I also would love to hear the ideas of @patrickdillon @makentenza  @dhellmann and @sdodson

Reply:
> Is it the idea with this logic of lookup of type offerings between all zones (regular and local) set the same instance between worker and edge?

No. I would expect all workers to be the same, and all edge to be the same, if edge and worker differ I think that's ok

> nodes between different edge groups may not share the same workload. For that reason, we are setting taints and custom labels by zoneGroup to give control to admins/developers to choose what workloads to run for each edge location. For example, if the administrator creates a new ingress (ALB) sharing nodes from different Local Zones zone groups (i.e.: LAS and LAX), their end-users will have different experiences (sometimes worst) in terms of network latency.

This is a fair point, I don't particularly have a strong opinion on this

Reply:
Yeah that's really interesting and helps argue the case for edge groups being completely independent of one another, if you can include a bit about this that would be cool, thank you!

Reply:
@JoelSpeed section added with a summary of the benchmark mentioned above - which led to the proposal of allowing different types on the edge pools from different zone groups:
https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R698-R769

Thanks!